GS/HW: Refactor StretchRect() to have single entry to renderer/reduce duplication.

TJnotJT opened the pull request:
### Description of Changes
Refactor some of the StretchRect() code to allow clearer separation between interface/implementation, reduce code duplicaton between renderers, make clearer which shaders have multiple pipline objects for the channel mask (DX12, Vulkan, Metal). Feedback/criticism is much appreciated.

### Rationale behind Changes
This allows some of the changes in the PR: https://github.com/PCSX2/pcsx2/pull/13133 to be made more easily. However, it also has the following benefits:
1. Make clearer the interface to StretchRect(). Rename the API-specific handling to DoStretchRect().
2. Have a single entry point from GSDevice::StretchRect() to subclasses so that changes are easier to make (like in the PR mentioned above). The entry point is the pure virtual member GSDevice::DoStretchRect().
3. Reduce duplication in renderers. Minor change, but put some of the parameter preprocessing/checking in GSDevice instead of API-specific subclasses.
4. Make it clearer in DX12/Vulkan/Metal that some shaders have multiple pipline objects for the channel mask and move a bit of duplicated code to GSDevice.

### Suggested Testing Steps
This affect pretty much all HW texture handling related so full dump runs in all renderers will be needed (will take some time). So far partial dump runs in GL/Vulkan/DX11/DX12 are done. Unfortunately, I am not able to compile or test Metal so would need some help with that.

Update: Full dumps runs in OpenGL, Vulkan, DX11, DX12 show no regressions.

### Did you use AI to help find, test, or implement this issue or feature?
GitHub Copilot.

lightningterror reviewed (commented):
Changes in TC are unnecessary.

TJnotJT commented:
> Changes in TC are unnecessary.

Thanks for catching, I reverted the change in the TC.

Also changed functions name `HasVariableChannelMask()` to `HasVariableWriteMask()` to be consistent with `ShaderConvertWriteMask()`.

Also a couple lines changed VK/DX12/MTL to use this `GetShaderIndexForMask()` instead of it being duplicated in each renderer.

TJnotJT commented:
Rebased with no merge conflicts.

TellowKrinkle commented on pcsx2/GS/Renderers/Common/GSDevice.h:
```diff
@@ -922,6 +955,11 @@ class GSDevice : public GSAlignedClass<32>
+protected:
+	// Entry point to the renderer-specific StretchRect code.
+	virtual void DoStretchRect(GSTexture* sTex, const GSVector4& sRect, GSTexture* dTex, const GSVector4& dRect,
+		GSHWDrawConfig::ColorMaskSelector cms, ShaderConvert shader = ShaderConvert::COPY, bool linear = true) = 0;
```
Defaulted parameters are nice for convenience on methods that get called a lot, but this only ever gets called from the three StretchRect implementations.  Doesn't seem worth it to me.

TJnotJT replied:
Yeah, removed the defaults here as well - they also weren't being used.

TellowKrinkle commented on pcsx2/GS/Renderers/Common/GSDevice.cpp:
```diff
@@ -723,7 +723,40 @@ GSTexture* GSDevice::CreateTexture(int w, int h, int mipmap_levels, GSTexture::F
+	pxAssert(linear ? SupportsBilinear(shader) : SupportsNearest(shader));
+	pxAssert(HasVariableWriteMask(shader) || cms.wrgba == ShaderConvertWriteMask(shader));
+	if (HasVariableWriteMask(shader))
+		GL_INS("ColorCopy Red:%d Green:%d Blue:%d Alpha:%d", cms.wr, cms.wg, cms.wb, cms.wa);
```
This will now be added every single time anyone uses a COPY shader.  Previously, it was only added if someone called the function that allows specifying a color mask.

TellowKrinkle commented on pcsx2/GS/Renderers/Common/GSDevice.cpp:
```diff
@@ -723,7 +723,40 @@ GSTexture* GSDevice::CreateTexture(int w, int h, int mipmap_levels, GSTexture::F
+{
+	pxAssert((dTex && dTex->IsDepthStencil()) == HasDepthOutput(shader));
+	pxAssert(linear ? SupportsBilinear(shader) : SupportsNearest(shader));
+	pxAssert(HasVariableWriteMask(shader) || cms.wrgba == ShaderConvertWriteMask(shader));
```
The function that allows a color mask should always `pxAssert(HasVariableWriteMask(shader))`, even if the color mask happens to be 0xf.  Other StretchRect calls shouldn't need to check this at all, since they don't allow specifying color masks.

TJnotJT commented on pcsx2/GS/Renderers/Common/GSDevice.h:
```diff
@@ -1037,9 +1075,10 @@ class GSDevice : public GSAlignedClass<32>
-	virtual void StretchRect(GSTexture* sTex, const GSVector4& sRect, GSTexture* dTex, const GSVector4& dRect, ShaderConvert shader = ShaderConvert::COPY, bool linear = true) = 0;
-	virtual void StretchRect(GSTexture* sTex, const GSVector4& sRect, GSTexture* dTex, const GSVector4& dRect, bool red, bool green, bool blue, bool alpha, ShaderConvert shader = ShaderConvert::COPY) = 0;
 
+	void StretchRect(GSTexture* sTex, const GSVector4& sRect, GSTexture* dTex, const GSVector4& dRect, GSHWDrawConfig::ColorMaskSelector cms, ShaderConvert shader = ShaderConvert::COPY, bool linear = true);
```
Sounds good - I renamed as such and removed the default parameters, which weren't being used anyway.
